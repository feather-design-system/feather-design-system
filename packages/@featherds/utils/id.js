const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
