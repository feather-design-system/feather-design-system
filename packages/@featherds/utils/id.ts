const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
