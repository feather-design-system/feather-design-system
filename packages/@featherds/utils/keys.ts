const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
