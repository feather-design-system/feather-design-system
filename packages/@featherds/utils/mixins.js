var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
