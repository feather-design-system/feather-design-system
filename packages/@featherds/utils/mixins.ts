const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
