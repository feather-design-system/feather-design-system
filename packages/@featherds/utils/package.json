{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
