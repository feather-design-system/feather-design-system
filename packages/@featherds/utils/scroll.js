var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
