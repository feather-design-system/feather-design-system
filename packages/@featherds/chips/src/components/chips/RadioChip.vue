<template>
  <Chip
    :disabled="disabled"
    :condensed="condensed"
    class="focus hover"
    :class="{ selected: checked }"
    role="radio"
    ref="input"
    :aria-checked="checked ? 'true' : 'false'"
    :aria-disabled="disabled"
    :aria-labelledby="labelId"
    :tabindex="tabindex"
    @click="click"
    :allow-click="!disabled"
    @blur="blur"
  >
    <PreIcon v-if="hasIcon"><slot name="icon" /></PreIcon>
    <Label :id="labelId"><slot /></Label>
  </Chip>
</template>
<script>
import Chip from "../base/Chip";
import Label from "../base/Label";
import PreIcon from "../base/PreIcon";
import { getSafeId } from "@featherds/utils/id";
import { computed, inject, ref } from "vue";
export default {
  props: {
    value: undefined,
    disabled: {
      type: Boolean,
      default: false,
    },
    condensed: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    const first = ref(false);
    const checked = ref(false);
    const labelId = computed(() => getSafeId("chip-label-id"));
    const tabindex = computed(() => {
      if (first.value) {
        return 0;
      }
      return checked.value ? 0 : -1;
    });

    const input = ref(null);
    const focus = () => {
      input.value.$el.focus();
    };
    //register
    const register = inject("register");
    const blur = inject("blur");
    const select = inject("select");

    const vm = {
      first,
      focus,
      disabled: props.disabled,
      value: props.value,
      checked,
    };
    register(vm);
    const click = () => {
      select(vm);
    };
    return {
      labelId,
      tabindex,
      first,
      blur,
      click,
      input,
      checked,
    };
  },
  computed: {
    hasIcon() {
      const hasIcon =
        this.$slots.icon &&
        this.$slots
          .icon()
          .findIndex(
            (o) =>
              (o.children && o.children.length !== 0) ||
              typeof o.type === "object"
          ) !== -1;
      return hasIcon;
    },
  },
  components: {
    Chip,
    Label,
    PreIcon,
  },
};
</script>
<style lang="scss" scoped>
@import "~@featherds/styles/themes/variables";
@import "~@featherds/styles/mixins/typography";
[role="radio"] {
  cursor: pointer;
}
</style>
