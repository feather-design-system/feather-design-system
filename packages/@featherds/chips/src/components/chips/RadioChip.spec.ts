import RadioChip from "./RadioChip.vue";
import { mount } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
import { nextTick } from "vue";
jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

const mockProvide = () => {
  const register = jest.fn();
  const select = jest.fn();
  const blur = jest.fn();
  return {
    provide: {
      register,
      select,
      blur,
    },
  };
};

const getWrapper = function (
  options: Record<string, unknown> = {},
  addIcon = false
) {
  const props = options.props ? (options.props as Object) : {};
  options.props = {
    ...props,
    value: "" as unknown,
  };
  options.slots = getSlots(addIcon).slots;

  return mount(RadioChip, options);
};
const getSlots = (withIcon = false) => {
  const slots = {
    slots: {
      default: "Test",
    },
  } as {
    slots: {
      default: string;
      icon?: { template: string };
    };
  };
  if (withIcon) {
    slots.slots.icon = { template: "<span>*</span>" };
  }
  return slots;
};

describe("RadioChip", () => {
  describe("radio button", () => {
    it("should call register", () => {
      const { provide } = mockProvide();
      getWrapper({ global: { provide } });
      expect(provide.register).toHaveBeenCalled();
    });
    it("should be focusable when first is set even if not checked", async () => {
      const wrapper = getWrapper({ global: mockProvide() });
      wrapper.vm.first = true;
      await nextTick();
      expect(wrapper.vm.tabindex).toBe(0);
    });
    it("should call blur", async () => {
      const { provide } = mockProvide();
      const wrapper = getWrapper({ global: { provide } });
      wrapper.vm.blur();
      await nextTick();
      expect(provide.blur).toHaveBeenCalled();
    });

    it("should call select from provide", async () => {
      const { provide } = mockProvide();
      const value = 1;
      const wrapper = getWrapper({
        props: {
          value,
        },
        global: { provide },
      });

      wrapper.vm.click();
      await nextTick();
      expect(provide.select).toHaveBeenCalled();
    });

    it("should select element on click", async () => {
      const value = 1;
      const { provide } = mockProvide();
      const wrapper = getWrapper({
        props: {
          value,
        },
        global: { provide },
      });

      await wrapper.find("[role='radio']").trigger("click");
      expect(provide.select).toHaveBeenCalled();
    });
    it("should set aria-checked based on data", async () => {
      const wrapper = getWrapper({
        global: mockProvide(),
      });
      wrapper.vm.checked = true;
      await nextTick();
      expect(wrapper.element).toMatchSnapshot();
      wrapper.vm.checked = false;
      await nextTick();
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should set aria-disabled based on property", async () => {
      const wrapper = getWrapper({
        props: {
          disabled: true,
        },
        global: mockProvide(),
      });
      expect(wrapper.element).toMatchSnapshot();
      await wrapper.setProps({
        disabled: false,
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  it("should render a standard chip", () => {
    const wrapper = getWrapper({ global: mockProvide() });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render a standard chip with icon", () => {
    const wrapper = getWrapper({ global: mockProvide() }, true);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render a disabled chip", () => {
    const props = {
      disabled: true,
    };
    const wrapper = getWrapper({ props, global: mockProvide() });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render condensed", () => {
    const props = {
      condensed: true,
    };
    const wrapper = getWrapper({ props, global: mockProvide() });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render condensed from a group", () => {
    const { provide: radioProvide } = mockProvide();
    const provide = {
      condensed: true,
      global: { provide: radioProvide },
    };
    const wrapper = getWrapper(provide);
    expect(wrapper.element).toMatchSnapshot();
  });
});
