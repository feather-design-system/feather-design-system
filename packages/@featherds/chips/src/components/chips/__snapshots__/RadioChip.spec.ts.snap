// Vitest Snapshot v1

exports[`RadioChip > radio button > should set aria-checked based on data 1`] = `
<div
  allow-click="true"
  aria-checked="true"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip focus hover selected"
  role="radio"
  tabindex="0"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > radio button > should set aria-checked based on data 2`] = `
<div
  allow-click="true"
  aria-checked="false"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > radio button > should set aria-disabled based on property 1`] = `
<div
  allow-click="false"
  aria-checked="false"
  aria-disabled="true"
  aria-labelledby="chip-label-id"
  class="chip disabled focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > radio button > should set aria-disabled based on property 2`] = `
<div
  allow-click="true"
  aria-checked="false"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > should render a disabled chip 1`] = `
<div
  allow-click="false"
  aria-checked="false"
  aria-disabled="true"
  aria-labelledby="chip-label-id"
  class="chip disabled focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > should render a standard chip 1`] = `
<div
  allow-click="true"
  aria-checked="false"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > should render a standard chip with icon 1`] = `
<div
  allow-click="true"
  aria-checked="false"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip focus hover"
  role="radio"
  tabindex="-1"
>
  
  <span
    class="chip-icon"
    role="presentation"
  >
    
    
    <span>
      *
    </span>
    
    
  </span>
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > should render condensed 1`] = `
<div
  allow-click="true"
  aria-checked="false"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip condensed focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;

exports[`RadioChip > should render condensed from a group 1`] = `
<div
  allow-click="true"
  aria-checked="false"
  aria-disabled="false"
  aria-labelledby="chip-label-id"
  class="chip focus hover"
  role="radio"
  tabindex="-1"
>
  
  <!--v-if-->
  <span
    class="label"
    id="chip-label-id"
  >
    
    
    Test
    
    
  </span>
  
  <!--v-if-->
</div>
`;
